File: app/src/main/java/com/star_zero/example/androidmvvm/application/TaskService.java
package com.star_zero.example.androidmvvm.application;

import com.star_zero.example.androidmvvm.application.dto.TaskDTO;
import com.star_zero.example.androidmvvm.domain.task.Task;
import com.star_zero.example.androidmvvm.domain.task.TaskRepository;
import com.star_zero.example.androidmvvm.domain.task.TaskValidator;

import java.util.List;

import javax.inject.Inject;

import rx.Observable;
import rx.Observer;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;
import rx.subjects.PublishSubject;
import rx.subscriptions.CompositeSubscription;
import timber.log.Timber;

public class TaskService {

    private TaskRepository taskRepository;

    private CompositeSubscription subscriptions = new CompositeSubscription();

    private TaskValidator validator = new TaskValidator();

    @Inject
    TaskService(TaskRepository taskRepository) {
        this.taskRepository = taskRepository;
    }

    public void getTasks() {
        subscriptions.add(taskRepository.getTasks()
                .subscribeOn(Schedulers.newThread())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Observer<List<Task>>() {
                    @Override
                    public void onCompleted() {
                    }

                    @Override
                    public void onError(Throwable e) {
                        Timber.w(e);
                        errorGetTasksSubject.onNext(null);
                    }

                    @Override
                    public void onNext(List<Task> tasks) {
                        tasksSubject.onNext(tasks);
                    }
                })
        );
    }

    public void save(TaskDTO taskDTO) {
        taskDTO.clearValidationErrors();

        if (!validator.validate(taskDTO.getTitle(), taskDTO.getDescription())) {
            validationErrorSubject.onNext(null);
            taskDTO.setValidationErrors(validator.getErrors());
            return;
        }

        if (taskDTO.getTask() == null) {
            saveNewTask(taskDTO);
        } else {
            updateTask(taskDTO);
        }
    }

    private void saveNewTask(TaskDTO taskDTO) {

        Task task = Task.createNewTask(taskRepository.generateTaskId(), taskDTO.getTitle(), taskDTO.getDescription());

        subscriptions.add(taskRepository.save(task)
                .subscribeOn(Schedulers.newThread())
                .unsubscribeOn(AndroidSchedulers.mainThread())
                .subscribe(new Observer<Boolean>() {
                    @Override
                    public void onCompleted() {
                    }

                    @Override
                    public void onError(Throwable e) {
                        Timber.w(e);
                        errorSaveTaskSubject.onNext(null);
                    }

                    @Override
                    public void onNext(Boolean result) {
                        if (result) {
                            successSaveTaskSubject.onNext(null);
                        } else {
                            errorSaveTaskSubject.onNext(null);
                        }
                    }
                })
        );
    }

    private void updateTask(TaskDTO taskDTO) {
        Task task = taskDTO.getTask();
        task.update(taskDTO.getTitle(), taskDTO.getDescription());

        subscriptions.add(taskRepository.update(task)
                .subscribe(new Observer<Boolean>() {
                    @Override
                    public void onCompleted() {
                    }

                    @Override
                    public void onError(Throwable e) {
                        Timber.w(e);
                        errorSaveTaskSubject.onNext(null);
                    }

                    @Override
                    public void onNext(Boolean result) {
                        if (result) {
                            successSaveTaskSubject.onNext(null);
                        } else {
                            errorSaveTaskSubject.onNext(null);
                        }
                    }
                })
        );

    }

    public void changeCompleteState(Task task, boolean completed) {
        if (completed) {
            task.completeTask();
        } else {
            task.activateTask();
        }

        subscriptions.add(taskRepository.update(task)
                .subscribe(new Observer<Boolean>() {
                    @Override
                    public void onCompleted() {
                    }

                    @Override
                    public void onError(Throwable e) {
                        Timber.w(e);
                        errorChangeCompleteStateSubject.onNext(null);

                        // rollback
                        if (completed) {
                            task.activateTask();
                        } else {
                            task.completeTask();
                        }
                    }

                    @Override
                    public void onNext(Boolean result) {
                    }
                })
        );
    }

    public void deleteTask(Task task) {

        taskRepository.delete(task).subscribe(new Observer<Boolean>() {
            @Override
            public void onCompleted() {
            }

            @Override
            public void onError(Throwable e) {
                Timber.w(e);
                errorDeleteTaskSubject.onNext(null);
            }

            @Override
            public void onNext(Boolean result) {
                if (result) {
                    successDeleteTaskSubject.onNext(null);
                } else {
                    errorDeleteTaskSubject.onNext(null);
                }
            }
        });
    }

    public void onDestroy() {
        subscriptions.clear();
    }

    // ----------------------
    // Notification
    // ----------------------

    // getTasks
    private final PublishSubject<List<Task>> tasksSubject = PublishSubject.create();
    public final Observable<List<Task>> tasks = tasksSubject.asObservable();
    private final PublishSubject<Void> errorGetTasksSubject = PublishSubject.create();
    public final Observable<Void> errorGetTasks = errorGetTasksSubject.asObservable();

    // save
    private final PublishSubject<Void> successSaveTaskSubject = PublishSubject.create();
    public final Observable<Void> successSaveTask = successSaveTaskSubject.asObservable();
    private final PublishSubject<Void> errorSaveTaskSubject = PublishSubject.create();
    public final Observable<Void> errorSaveTask = errorSaveTaskSubject.asObservable();

    // validation
    private final PublishSubject<Void> validationErrorSubject = PublishSubject.create();
    public final Observable<Void> validationError = validationErrorSubject.asObservable();

    // completeTask, activateTask
    private final PublishSubject<Void> errorChangeCompleteStateSubject = PublishSubject.create();
    public final Observable<Void> errorChangeCompleteState = errorChangeCompleteStateSubject.asObservable();

    // delete
    private final PublishSubject<Void> successDeleteTaskSubject = PublishSubject.create();
    public final Observable<Void> successDeleteTask = successDeleteTaskSubject.asObservable();
    private final PublishSubject<Void> errorDeleteTaskSubject = PublishSubject.create();
    public final Observable<Void> errorDeleteTask = errorDeleteTaskSubject.asObservable();
}

File: app/src/main/kotlin/com/star_zero/example/androidmvvm/application/TaskService.kt
package com.star_zero.example.androidmvvm.application

import android.arch.lifecycle.LifecycleOwner
import android.arch.lifecycle.LiveData
import com.star_zero.example.androidmvvm.application.dto.TaskDTO
import com.star_zero.example.androidmvvm.domain.task.Task
import com.star_zero.example.androidmvvm.domain.task.TaskRepository
import com.star_zero.example.androidmvvm.domain.task.TaskValidator
import com.star_zero.example.androidmvvm.utils.AsyncResult
import com.star_zero.example.androidmvvm.utils.PublishLiveData
import com.star_zero.example.androidmvvm.utils.extension.fire
import com.star_zero.example.androidmvvm.utils.extension.observe
import timber.log.Timber
import javax.inject.Inject

class TaskService @Inject constructor(val taskRepository: TaskRepository) {

    private val validator = TaskValidator()

    fun fetchTasks(owner: LifecycleOwner) {
        taskRepository.fetchTasks().observe(owner) {
            when (it) {
                is AsyncResult.Success -> {
                    tasks.value = it.value
                }
                is AsyncResult.Error -> {
                    Timber.w(it.error)
                    errorFetchTasks.fire()
                }
            }
        }
    }

    fun save(taskDTO: TaskDTO, owner: LifecycleOwner) {
        taskDTO.clearValidationErrors()

        if (!validator.validate(taskDTO.title, taskDTO.description)) {
            validationError.fire()
            taskDTO.setValidationErrors(validator.errors)
            return
        }

        if (taskDTO.task == null) {
            saveNewTask(taskDTO, owner)
        } else {
            updateTask(taskDTO, owner)
        }
    }

    private fun saveNewTask(taskDTO: TaskDTO, owner: LifecycleOwner) {
        val task = Task.createNewTask(taskRepository.generateTaskId(), taskDTO.title!!, taskDTO.description)

        taskRepository.save(task).observe(owner) {
            when(it) {
                is AsyncResult.Success -> successSaveTask.fire()
                is AsyncResult.Error -> {
                    Timber.w(it.error)
                    errorSaveTask.fire()
                }
            }
        }
    }

    private fun updateTask(taskDTO: TaskDTO, owner: LifecycleOwner) {
        val task = taskDTO.task!!
        task.update(taskDTO.title!!, taskDTO.description)

        taskRepository.update(task).observe(owner) {
            when(it) {
                is AsyncResult.Success -> successSaveTask.fire()
                is AsyncResult.Error -> {
                    Timber.w(it.error)
                    errorSaveTask.fire()
                }
            }
        }
    }

    fun changeCompleteState(task: Task, completed: Boolean, owner: LifecycleOwner) {
        if (completed) {
            task.completeTask()
        } else {
            task.activateTask()
        }

        taskRepository.update(task).observe(owner) {
            when(it) {
                is AsyncResult.Error -> {
                    Timber.w(it.error)
                    errorChangeCompleteState.fire()

                    // rollback
                    if (completed) {
                        task.activateTask()
                    } else {
                        task.completeTask()
                    }
                }
            }
        }
    }

    fun deleteTask(task: Task, owner: LifecycleOwner) {
        taskRepository.delete(task).observe(owner) {
            when (it) {
                is AsyncResult.Success -> successDeleteTask.fire()
                is AsyncResult.Error -> {
                    Timber.w(it.error)
                    errorDeleteTask.fire()
                }
            }
        }
    }

    // fetchTasks

    private val tasks = PublishLiveData<List<Task>?>()
    fun tasks(): LiveData<List<Task>?> = tasks
    private val errorFetchTasks = PublishLiveData<Void>()
    fun errorFetchTasks(): LiveData<Void> = errorFetchTasks

    // save

    private val successSaveTask = PublishLiveData<Void>()
    fun successSaveTask(): LiveData<Void> = successSaveTask
    private val errorSaveTask = PublishLiveData<Void>()
    fun errorSaveTask(): LiveData<Void> = errorSaveTask

    // validation

    private val validationError = PublishLiveData<Void>()
    fun validationError(): LiveData<Void> = validationError

    // completeTask, activateTask

    private val errorChangeCompleteState = PublishLiveData<Void>()
    fun errorChangeCompleteState(): LiveData<Void> = errorChangeCompleteState

    // delete

    private val successDeleteTask = PublishLiveData<Void>()
    fun successDeleteTask(): LiveData<Void> = successDeleteTask
    private val errorDeleteTask = PublishLiveData<Void>()
    fun errorDeleteTask(): LiveData<Void> = errorDeleteTask

}